File: app/src/main/java/com/megablok10/app/breach/LootCrypto.kt
package com.megablok10.app.breach

import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Шифрует payload лут-слотов AES-GCM ключом этой игры (см. [deriveKey]).
 * Модель угрозы — игрок с телефоном и обычным сканером QR на полигоне, не
 * реверс-инженер: цель не защититься от разбора APK, а не дать прочитать
 * шард стандартным QR-сканером мимо мини-игры взлома. Ключ намеренно не
 * Android Keystore — его не нужно привязывать к устройству, все телефоны
 * этой игры должны уметь читать один и тот же контейнер.
 */
object LootCrypto {
    // 32 байта — AES-256, ключ по умолчанию, если GAME_SECRET не настроен
    // (см. deriveKey) — тот самый ключ, что был единственным в v1. Смена
    // этой константы делает нечитаемыми payload'ы всех уже напечатанных
    // QR игр без GAME_SECRET.
    private val FALLBACK_KEY = byteArrayOf(
        0x4d, 0x42, 0x31, 0x30, 0x2d, 0x4c, 0x4f, 0x4f,
        0x54, 0x2d, 0x76, 0x31, 0x2d, 0x6b, 0x65, 0x79,
        0x2d, 0x62, 0x72, 0x65, 0x61, 0x63, 0x68, 0x2d,
        0x70, 0x72, 0x6f, 0x74, 0x6f, 0x63, 0x6f, 0x6c
    )
    // "Нет соли" по RFC 5869 — HashLen (32) нулевых байт, явно, а не по умолчанию.
    private val HKDF_SALT = ByteArray(32)
    private val HKDF_INFO = "mb10-loot-v1".toByteArray(Charsets.UTF_8)
    private const val IV_BYTES = 12
    private const val TAG_BITS = 128

    /**
     * Ключ AES-256 для лута этой игры: без GAME_SECRET — FALLBACK_KEY (как
     * было всегда), с GAME_SECRET — HKDF-SHA256(secret) с тем же salt/info,
     * что и зеркало admin-web/server/src/lib/lootCrypto.ts (deriveLootKey),
     * так что ключ совпадает у всех телефонов этой игры, но не совпадает с
     * ключом другой игры или со старым публичным ключом. secret приходит из
     * CollectorSettings.gameSecret — доставлен полем QR персонажа (см.
     * docs/provisioning-qr.md), отдельной инфраструктуры не требуется.
     */
    fun deriveKey(gameSecret: String?): ByteArray {
        val secret = gameSecret?.trim()?.takeIf { it.isNotEmpty() } ?: return FALLBACK_KEY
        return hkdfSha256(ikm = secret.toByteArray(Charsets.UTF_8), salt = HKDF_SALT, info = HKDF_INFO, length = 32)
    }

    fun encrypt(plain: String, key: ByteArray): String {
        val iv = ByteArray(IV_BYTES).also { java.security.SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BITS, iv))
        val encrypted = cipher.doFinal(plain.toByteArray(Charsets.UTF_8))
        return Base64.getEncoder().encodeToString(iv + encrypted)
    }

    /** null — payload битый или зашифрован другим ключом (например, QR другой игры или прошлого акта). */
    fun decrypt(payloadB64: String, key: ByteArray): String? = try {
        val bytes = Base64.getDecoder().decode(payloadB64)
        val iv = bytes.copyOfRange(0, IV_BYTES)
        val body = bytes.copyOfRange(IV_BYTES, bytes.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BITS, iv))
        String(cipher.doFinal(body), Charsets.UTF_8)
    } catch (e: Exception) {
        null
    }

    /** HKDF-SHA256 (RFC 5869), только Extract + один блок Expand — этого хватает ровно на length <= 32. */
    private fun hkdfSha256(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        require(length in 1..32) { "single-block HKDF-Expand only covers up to 32 bytes" }
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(salt, "HmacSHA256"))
        val prk = mac.doFinal(ikm)
        mac.init(SecretKeySpec(prk, "HmacSHA256"))
        val t1 = mac.doFinal(info + byteArrayOf(0x01))
        return t1.copyOf(length)
    }
}

File: admin-web/server/src/lib/lootCrypto.ts
import { createCipheriv, createDecipheriv, createHmac, randomBytes } from "node:crypto";

/**
 * Зеркало app/src/main/java/com/megablok10/app/breach/LootCrypto.kt — тот
 * же ключ (см. deriveLootKey), тот же AES-256-GCM. Java's Cipher.doFinal()
 * в GCM-режиме кладёт тег аутентификации (16 байт) СРАЗУ ЗА шифротекстом в
 * один массив; Node отдаёт тег отдельным вызовом getAuthTag() — здесь
 * вручную склеиваем в тот же порядок байт (iv || ciphertext || tag), чтобы
 * Android читал ровно то же, что генерирует этот сервер, и наоборот.
 */
// Ключ по умолчанию, если GAME_SECRET не настроен (см. deriveLootKey) — тот
// самый ключ, что был единственным в v1. Смена этой константы делает
// нечитаемыми payload'ы всех уже напечатанных QR игр без GAME_SECRET.
const FALLBACK_KEY = Buffer.from("MB10-LOOT-v1-key-breach-protocol", "utf8"); // 32 байт — AES-256
const HKDF_SALT = Buffer.alloc(32); // "нет соли" по RFC 5869 — HashLen нулей, явно, а не по умолчанию
const HKDF_INFO = Buffer.from("mb10-loot-v1", "utf8");
const IV_BYTES = 12;
const TAG_BYTES = 16;

/**
 * Ключ AES-256 для лута этой игры: без GAME_SECRET — FALLBACK_KEY (как было
 * всегда), с GAME_SECRET — HKDF-SHA256(secret) с тем же salt/info на Android
 * (LootCrypto.deriveKey), так что ключ у всех телефонов этой игры совпадает,
 * но не совпадает с ключом другой игры или со старым публичным ключом.
 * До GAME_SECRET ключ был один статичный и лежал прямо в этом файле в
 * открытом репозитории — расшифровать любой QR любой игры мог кто угодно,
 * не только реверс-инженер APK. HKDF от секрета конкретной игры поднимает
 * это до «нужно знать секрет именно этой игры», без нового формата QR и
 * новой инфраструктуры — secret уже доставляется на телефон полем QR
 * персонажа (см. docs/provisioning-qr.md).
 */
export function deriveLootKey(gameSecret: string | undefined): Buffer {
  const secret = gameSecret?.trim();
  if (!secret) return FALLBACK_KEY;
  const prk = createHmac("sha256", HKDF_SALT).update(secret, "utf8").digest();
  return createHmac("sha256", prk).update(Buffer.concat([HKDF_INFO, Buffer.from([0x01])])).digest();
}

export function encryptLoot(plain: string, key: Buffer): string {
  const iv = randomBytes(IV_BYTES);
  const cipher = createCipheriv("aes-256-gcm", key, iv, { authTagLength: TAG_BYTES });
  const ciphertext = Buffer.concat([cipher.update(plain, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  return Buffer.concat([iv, ciphertext, tag]).toString("base64");
}

/** null — payload битый или зашифрован другим ключом (совместимо с LootCrypto.decrypt на клиенте). */
export function decryptLoot(payloadB64: string, key: Buffer): string | null {
  try {
    const bytes = Buffer.from(payloadB64, "base64");
    const iv = bytes.subarray(0, IV_BYTES);
    const tag = bytes.subarray(bytes.length - TAG_BYTES);
    const ciphertext = bytes.subarray(IV_BYTES, bytes.length - TAG_BYTES);
    const decipher = createDecipheriv("aes-256-gcm", key, iv, { authTagLength: TAG_BYTES });
    decipher.setAuthTag(tag);
    return Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString("utf8");
  } catch {
    return null;
  }
}
